src/FindIntrinsics.cpp

```diff
@@ -639,33 +642,103 @@ Expr lower_saturating_sub(const Expr &a, const Expr &b) {
 Expr lower_halving_add(const Expr &a, const Expr &b) {
     internal_assert(a.type() == b.type());
     // Borrowed from http://aggregate.org/MAGIC/#Average%20of%20Integers
-    return (a & b) + make_shift_right((a ^ b), 1);
+    return (a & b) + ((a ^ b) >> 1);
 }
 
 Expr lower_halving_sub(const Expr &a, const Expr &b) {
     internal_assert(a.type() == b.type());
-    return make_shift_right(a, 1) - make_shift_right(b, 1) - make_shift_right((b & 1) - (a & 1) + 1, 1);
+    return (a >> 1) - (b >> 1) - (((b & 1) - (a & 1) + 1) >> 1);
 }
 
 // TODO: These should using rounding_shift_right, but lowering that
 // results in double widening and the simplifier doesn't fix it.
 Expr lower_rounding_halving_add(const Expr &a, const Expr &b) {
     internal_assert(a.type() == b.type());
-    return make_shift_right(a, 1) + make_shift_right(b, 1) + make_shift_right((a & 1) + (b & 1) + 1, 1);
+    return (a >> 1) + (b >> 1) + (((a & 1) + (b & 1) + 1) >> 1);
 }
 
 Expr lower_rounding_halving_sub(const Expr &a, const Expr &b) {
     internal_assert(a.type() == b.type());
-    return make_shift_right(a, 1) - make_shift_right(b, 1) + make_shift_right((a & 1) - (b & 1) + 1, 1);
-}
-
-Expr lower_mulhi_shr(const Type &result_type, const Expr &a, const Expr &b, const Expr &shift) {
-    return cast(result_type, widening_mul(a, b) >> simplify(shift + result_type.bits()));
+    return (a >> 1) - (b >> 1) + (((a & 1) - (b & 1) + 1) >> 1);
 }
 
 Expr lower_sorted_avg(const Expr &a, const Expr &b) {
     // b > a, so the following works without widening.
-    return a + (b - a) / 2;
+    return a + ((b - a) >> 1);
+}
+
+Expr lower_mul_shift_right(const Expr &a, const Expr &b, const Expr &q) {
+    internal_assert(a.type() == b.type());
+    int full_q = a.type().bits();
+    if (a.type().is_int()) {
+        full_q -= 1;
+    }
+    if (can_prove(q < full_q)) {
+        // Try to rewrite this to a "full precision" multiply by multiplying
+        // one of the operands and the denominator by a constant. We only do this
+        // if it isn't already full precision. This avoids infinite loops despite
+        // "lowering" this to another mul_shift_right operation.
+        Expr missing_q = full_q - q;
+        internal_assert(missing_q.type().bits() == b.type().bits());
+        Expr new_b = simplify(b << missing_q);
+        if (is_const(new_b) && can_prove(new_b >> missing_q == b)) {
+            return mul_shift_right(a, new_b, full_q);
+        }
+        Expr new_a = simplify(a << missing_q);
+        if (is_const(new_a) && can_prove(new_a >> missing_q == a)) {
```

Comment:
What's the can_prove guarding against? Overflow?

Reply:
Yes, this is just computing (a << missing_q) >> missing_q. If that isn't equal to a, then this rewrite isn't safe.